### PR TITLE
DAOS-10340:  support for additional container prop

### PR DESCRIPTION
**Pull request by ryon-jensen**
Addition of DAOS_PROP_CO_SCRUBBER_DISABLED property

 Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

### CODE REVIEW

File: Makefile
Comment (brianjmurrell) [inline Makefile:4]: It's a real pity that it's been almost 2 years since a new release has been made by upstream.

**Review by brianjmurrell — COMMENTED** [review]
Looks like a (pretty serious?) security issue from upstream is being reported by `rpmlint`:

[2024-01-04T16:12:02.990Z] mpifileutils-mpich.x86_64: E: missing-call-to-setgroups-before-setuid /usr/lib64/mpi/gcc/mpich/bin/dcp
[2024-01-04T16:12:02.990Z] mpifileutils-openmpi3.x86_64: E: missing-call-to-setgroups-before-setuid /usr/lib64/mpi/gcc/openmpi3/bin/dcp

I've opened hpc/mpifileutils#567 upstream about that.

File: mpifileutils.rpmlintrc
Comment (brianjmurrell) [inline mpifileutils.rpmlintrc:5]: Ahhh.  This is not quite ready to be removed yet as the [fix](https://github.com/rpm-software-management/rpmlint/commit/1056f04337a1b0e6e37d2fc29bdb500bc35b8fd7) for rpm-software-management/rpmlint#73 landed in _rpmlint_ 2.5.0:
image
and we (rather, Fedora) seems to be still using 2.4.0:

```
[2024-01-10T17:15:03.597Z] rpmlint: 2.4.0
```

I've [requested an update to 2.5.0 in Fedora 39](https://bugzilla.redhat.com/show_bug.cgi?id=2257744).

File: mpifileutils.rpmlintrc
Comment (brianjmurrell) [inline mpifileutils.rpmlintrc:8]: It's interesting that this is actually working with the missing RE escapes.  But we should also prevent it from matching, say, `/usr/lib64/openmpi/bin/dcpfoo':

```suggestion
addFilter(r'^mpifileutils-.+\.x86_64: E: missing-call-to-setgroups-before-setuid \/usr\/lib64\/.+\/bin\/dcp$')
```

File: mpifileutils.spec
Comment (brianjmurrell) [inline mpifileutils.spec:279]: Let's update the datestamp here when you next push.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: 0.11.1..3975d8483c2471c59c599d87516c7cea24bcc032.patch
Comment (daltonbohning) [inline 0.11.1..3975d8483c2471c59c599d87516c7cea24bcc032.patch:1]: Need to remove this unused, right?

File: mpifileutils.rpmlintrc
Comment (daltonbohning) [inline mpifileutils.rpmlintrc:8]: I think this is okay since we don't have 100% control over the source code

File: mpifileutils.spec
Comment (daltonbohning) [inline mpifileutils.spec:281]: Reminder that we need to update this to reference the final commit hash

**Review by daltonbohning — CHANGES_REQUESTED** [review]
LGTM - just need to update to the final commit hash after upstream PR lands

**Comment by daltonbohning** [comment]
@ryon-jensen I think the datamover test failure here is unrelated:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/ci-daos-stack-mpifileutils-pkg-PR-36-master/7/pipeline/

That test times out during the first IOR - before running any datamover or mpifileutils commands. And looking at the test, I see obvious improvements that could be made to ix the issue.
I created https://daosio.atlassian.net/browse/DAOS-15263

**Review by daltonbohning — APPROVED** [review]

File: mpifileutils.rpmlintrc
Comment (brianjmurrell) [reply mpifileutils.rpmlintrc:8]: Ideall, we should be filing a ticket upstream to get this fixed there.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]
Looks like you [need a `packaging/` update](https://jenkins-1.wolf.hpdd.intel.com:9090/blue/organizations/jenkins/daos-stack%2Fmpifileutils-pkg/detail/PR-36/9/pipeline/86).

I'd like to see if you get an rpmlint update in the builder when you push the above.

**Comment by brianjmurrell** [comment]
:roll_eyes: 
More [known failing tests](https://jenkins-1.wolf.hpdd.intel.com:9090/blue/organizations/jenkins/daos-stack%2Fdaos/detail/ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4/1/tests/) that have not been added to the skip list.

This PR is going to need to be force landed and the downstream branch [ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4](https://github.com/daos-stack/daos/tree/ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4) is going to need to be manually cleaned up once that has been done.

File: mpifileutils.rpmlintrc
Comment (daltonbohning) [reply mpifileutils.rpmlintrc]: If need be, I could push a fix myself and have it reviewed and merged by the repo maintainer

**Comment by daltonbohning** [comment]
> 🙄 More [known failing tests](https://jenkins-1.wolf.hpdd.intel.com:9090/blue/organizations/jenkins/daos-stack%2Fdaos/detail/ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4/1/tests/) that have not been added to the skip list.
> 
> This PR is going to need to be force landed and the downstream branch [ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4](https://github.com/daos-stack/daos/tree/ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4) is going to need to be manually cleaned up once that has been done.
> 
> @ryon-jensen I think we are going to have to start asking people working on dependency PRs to add known-failing-tests that have not been added to the skiplist to their PRs as a "do not run" (i.e. a negative test tag and adding ticket numbers for those tests to their commit messages) so that we can put an end the continual force-landing (and subsequent manual cleanup that results from it) that seems to be required for these dependency PRs.

FWIW, both of these are new failures for tests that have been passing for quite some time.
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4/1/pipeline

**Comment by ryon-jensen** [comment]
> > 🙄 More [known failing tests](https://jenkins-1.wolf.hpdd.intel.com:9090/blue/organizations/jenkins/daos-stack%2Fdaos/detail/ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4/1/tests/) that have not been added to the skip list.
> > This PR is going to need to be force landed and the downstream branch [ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4](https://github.com/daos-stack/daos/tree/ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4) is going to need to be manually cleaned up once that has been done.
> > @ryon-jensen I think we are going to have to start asking people working on dependency PRs to add known-failing-tests that have not been added to the skiplist to their PRs as a "do not run" (i.e. a negative test tag and adding ticket numbers for those tests to their commit messages) so that we can put an end the continual force-landing (and subsequent manual cleanup that results from it) that seems to be required for these dependency PRs.
> 
> FWIW, both of these are new failures for tests that have been passing for quite some time. https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4/1/pipeline But the failures in no way appear related to this packaging update. (unless I'm missing something)

I can't see how the failures can be related to this update. Can we go ahead and land it?

**Comment by daltonbohning** [comment]
Looking at the failures:
`datamover/large_file.py`:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4/1/artifact/Functional%20Hardware%20Medium/datamover/large_file.py/job.log
```
2024-02-27 05:46:07,467 process          L0416 DEBUG| [stdout] [2024-02-27T05:46:07] Copied 17.613 GiB (88%) in 20.014 secs (901.156 MiB/s) 3 secs left ...
2024-02-27 05:46:26,939 process          L0416 DEBUG| [stdout] [2024-02-27T05:46:26] [0] [/builddir/build/BUILD/mpifileutils-0.11.1/src/common/mfu_flist_copy.c:1766] ERROR: Read error when copying from `/testFile' to `/var/tmp/avocado_qpzmmrmb/avocado_job_0wbfs5cx/1-._datamover_large_file.py_DmvrPosixLargeFile.test_dm_large_file_dcp_run-container-datamover-dcp-client_processes-hosts-ior-client_ppn-pool-server_config-engines-0-storage-0-1-0090/test_dm_large_file_dcp0/testFile' (errno=2 No such file or directory)
2024-02-27 05:47:04,877 process          L0416 DEBUG| [stdout] [2024-02-27T05:47:04] [0] [/builddir/build/BUILD/mpifileutils-0.11.1/src/common/mfu_flist_copy.c:2237] ERROR: Failed to copy `/testFile' to `/var/tmp/avocado_qpzmmrmb/avocado_job_0wbfs5cx/1-._datamover_large_file.py_DmvrPosixLargeFile.test_dm_large_file_dcp_run-container-datamover-dcp-client_processes-hosts-ior-client_ppn-pool-server_config-engines-0-storage-0-1-0090/test_dm_large_file_dcp0/testFile'
2024-02-27 05:47:04,877 process          L0416 DEBUG| [stdout] [2024-02-27T05:47:04] Copied 18.195 GiB (91%) in 77.425 secs (240.644 MiB/s) done
```
So a 20G file was actively being copied from DAOS to `/var/tmp...` and suddenly got `No such file or directory`, which seems to be an error on the `/var/tmp` filesystem, not from DAOS


And the other failure
`datamover/negative_space.py`
https://daosio.atlassian.net/browse/DAOS-15263
Stems from the same filesystem `/var/tmp...` being very slow and the test timing out.


So both of these failures look system-related

File: mpifileutils.rpmlintrc
Comment (brianjmurrell) [reply mpifileutils.rpmlintrc]: Well, on the one hand, it would be nice to get this fixed as it just makes our environment more secure by doing so, even though it's just a test environment.  But given that we do not distribute this package I don't think it falls under an SDL requirement.  Perhaps an MSS, but I don't think MSS gets as detailed as that.

So, would be nice to have fixed, but is not a blocker of any sort for us.

Comment (daltonbohning) [reply mpifileutils.rpmlintrc]: I do have another mpifileutils update coming, so I can fix this upstream before I update this package again

**Comment by brianjmurrell** [comment]
> ```
> 2024-02-27 05:46:07,467 process          L0416 DEBUG| [stdout] [2024-02-27T05:46:07] Copied 17.613 GiB (88%) in 20.014 secs (901.156 MiB/s) 3 secs left ...
> 2024-02-27 05:46:26,939 process          L0416 DEBUG| [stdout] [2024-02-27T05:46:26] [0] [/builddir/build/BUILD/mpifileutils-0.11.1/src/common/mfu_flist_copy.c:1766] ERROR: Read error when copying from `/testFile' to `/var/tmp/avocado_qpzmmrmb/avocado_job_0wbfs5cx/1-._datamover_large_file.py_DmvrPosixLargeFile.test_dm_large_file_dcp_run-container-datamover-dcp-client_processes-hosts-ior-client_ppn-pool-server_config-engines-0-storage-0-1-0090/test_dm_large_file_dcp0/testFile' (errno=2 No such file or directory)
> ```

That's a **read** error.  While it's entirely possible that the `read()`ing is being done on the target (`/var/tmp/`) it seems much(!) more probably to me that the reading is happening on the source (`/testFile`).  Where is that file?  On DAOS?

> ```
> 2024-02-27 05:47:04,877 process          L0416 DEBUG| [stdout] [2024-02-27T05:47:04] [0] [/builddir/build/BUILD/mpifileutils-0.11.1/src/common/mfu_flist_copy.c:2237] ERROR: Failed to copy `/testFile' to `/var/tmp/avocado_qpzmmrmb/avocado_job_0wbfs5cx/1-._datamover_large_file.py_DmvrPosixLargeFile.test_dm_large_file_dcp_run-container-datamover-dcp-client_processes-hosts-ior-client_ppn-pool-server_config-engines-0-storage-0-1-0090/test_dm_large_file_dcp0/testFile'
> 2024-02-27 05:47:04,877 process          L0416 DEBUG| [stdout] [2024-02-27T05:47:04] Copied 18.195 GiB (91%) in 77.425 secs (240.644 MiB/s) done
> ```
> 
> So a 20G file was actively being copied from DAOS to `/var/tmp...` and suddenly got `No such file or directory`, which seems to be an error on the `/var/tmp` filesystem, not from DAOS

As above, I would disagree and say that the probability is higher that it's on where ever the source file (`/testFile`) is.
 
> And the other failure `datamover/negative_space.py` [daosio.atlassian.net/browse/DAOS-15263](https://daosio.atlassian.net/browse/DAOS-15263) Stems from the same filesystem `/var/tmp...` being very slow and the test timing out.

How do you know it's `/var/tmp` that's being slow?
 
> So both of these failures look system-related

But they don't present outside of this PR, correct?  If so, then we'd be adding more failures (regardless of their origin) by landing it, right?

**Comment by ryon-jensen** [comment]
> But they don't present outside of this PR, correct? If so, then we'd be adding more failures (regardless of their origin) by landing it, right?

I'll see if I can reproduce it.

**Comment by daltonbohning** [comment]
> That's a **read** error. While it's entirely possible that the `read()`ing is being done on the target (`/var/tmp/`) it seems much(!) more probably to me that the reading is happening on the source (`/testFile`). Where is that file? On DAOS?

Yes, you're right! Sorry, I missed that.

> As above, I would disagree and say that the probability is higher that it's on where ever the source file (`/testFile`) is.

Right. We need to verify.

> How do you know it's `/var/tmp` that's being slow?

Sorry, I mixed up the logs here. For this one, it's `/mnt/share/`. It's been a general problem with the datamover tests when they try to write multi-GB files to that FS.
And from this log
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/ci-daos-stack-mpifileutils-pkg-PR-36-release-2.4/1/artifact/Functional%20Hardware%20Medium/datamover/negative_space.py/job.log
we use IOR to create the file in `/mnt/share/` and it took 2 minutes just to create the file
```
2024-02-27 06:00:19,451 process          L0416 DEBUG| [stdout] Commencing write performance test: Tue Feb 27 06:00:19 2024
2024-02-27 06:02:16,239 process          L0416 DEBUG| [stdout] write     52.61      2451.69    0.000408    6291456    1024.00    0.002332   2.51       114.28     116.79     0   
2024-02-27 06:02:16,239 process          L0416 DEBUG| [stdout] Max Write: 52.61 MiB/sec (55.16 MB/sec)
2024-02-27 06:02:16,239 process          L0416 DEBUG| [stdout] 
2024-02-27 06:02:16,239 process          L0416 DEBUG| [stdout] Summary of all tests:
2024-02-27 06:02:16,239 process          L0416 DEBUG| [stdout] Operation   Max(MiB)   Min(MiB)  Mean(MiB)     StdDev   Max(OPs)   Min(OPs)  Mean(OPs)     StdDev    Mean(s) Stonewall(s) Stonewall(MiB)
2024-02-27 06:02:16,239 process          L0416 DEBUG| [stdout]  Test# #Tasks tPN reps fPP reord reordoff reordrand seed segcnt   blksiz    xsize aggs(MiB)   API RefNum
2024-02-27 06:02:16,239 process          L0416 DEBUG| [stdout] write          52.61      52.61      52.61       0.00      52.61      52.61      52.61       0.00  116.78812         NA            NA     0      1   1    1   0     0        1         0    0      1 6442450944  1048576    6144.0 POSIX      0
2024-02-27 06:02:16,239 process          L0416 DEBUG| [stdout] Finished            : Tue Feb 27 06:02:16 2024
```
And the test only has a timeout of 3 minutes.


> But they don't present outside of this PR, correct? If so, then we'd be adding more failures (regardless of their origin) by landing it, right?

I agree. We need to verify.

**Comment by daltonbohning** [comment]
> > But they don't present outside of this PR, correct? If so, then we'd be adding more failures (regardless of their origin) by landing it, right?
> 
> I'll see if I can reproduce it.

Let me know if you want help with this

**Comment by brianjmurrell** [comment]
> Sorry, I mixed up the logs here. For this one, it's `/mnt/share/`. It's been a general problem with the datamover tests when they try to write multi-GB files to that FS.

`/mnt/share/` is a(n NFS) shared filesystem for the test cluster.  It's definitely going to be slower than a local filesystem.  On VM clusters, it would be running over the virtualized network, which should be pretty performant since effectively it's all local within a single machine.  On hardware nodes, it would be running over a real network and so be limited by the physical network.  I don't know if we are using Ethernet for that or IB, or something else.

Why is a test writing to that filesystem anyway?  How does that test DAOS?  That space was really only ever intended to be a gathering place for test results, logging, etc. and coordinating tests.

If some kind of performant shared space is needed to actually run tests on, that probably needs a separate requirements specification with performance expectations so that it can be properly resourced and provisioned.

**Comment by daltonbohning** [comment]
> `/mnt/share/` is a(n NFS) shared filesystem for the test cluster. It's definitely going to be slower than a local filesystem. On VM clusters, it would be running over the virtualized network, which should be pretty performant since effectively it's all local within a single machine. 

I don't have data for this, but some datamover tests were specifically moved from VM to HW because the /mnt/share on VMs would sporadically drop to a trickle of KBs.

> On hardware nodes, it would be running over a real network and so be limited by the physical network. I don't know if we are using Ethernet for that or IB, or something else.

I haven't checked, but in general this FS on HW has been stable/quick enough for the few datamover tests using it to not have issues. Or maybe they just aren't using it as much anymore, I would have to check the tests.

> Why is a test writing to that filesystem anyway? How does that test DAOS? That space was really only ever intended to be a gathering place for test results, logging, etc. and coordinating tests.
> 
> If some kind of performant shared space is needed to actually run tests on, that probably needs a separate requirements specification with performance expectations so that it can be properly resourced and provisioned.

The tests are for datamover. For example, copying from DAOS to a POSIX (non-dfuse) FS. Before my time, I think it was hoped to have Lustre, but that was never realized.

**Comment by daltonbohning** [comment]
I saw the `datamover/negative_space.py` (DAOS-15263) fail the same way on a different PR:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13210/27/artifact/Functional%20Hardware%20Medium/datamover/negative_space.py/results.html

So I think we can say that one is pre-existing.

**Comment by brianjmurrell** [comment]
> The tests are for datamover. For example, copying from DAOS to a POSIX (non-dfuse) FS.

Does it need to be a shared filesystem on the destination of the copy?  Would a local filesystem on a single test node not suffice if it's just to test a copy to a POSIX filesystem?

**Comment by brianjmurrell** [comment]
> > The tests are for datamover. For example, copying from DAOS to a POSIX (non-dfuse) FS.
> 
> Does it need to be a shared filesystem on the destination of the copy? Would a local filesystem on a single test node not suffice if it's just to test a copy to a POSIX filesystem?

Can we have it skipped so that it doesn't cause this much effort and disruption in the future?  This really is a huge waste of engineering time and velocity to have to do this sort of thing with every dependency update that happens.

**Comment by daltonbohning** [comment]
> > The tests are for datamover. For example, copying from DAOS to a POSIX (non-dfuse) FS.
> 
> Does it need to be a shared filesystem on the destination of the copy? Would a local filesystem on a single test node not suffice if it's just to test a copy to a POSIX filesystem?

Switching to use the local FS has been a solution for some other tests, and might be for these too

**Comment by daltonbohning** [comment]
> > So I think we can say that one is pre-existing.
> 
> Can we have it skipped so that it doesn't cause this much effort and disruption in the future? This really is a huge waste of engineering time and velocity to have to do this sort of thing with every dependency update that happens.

I'm actively working on a workaround now. I agree it's a disruption to deal with known failures, but this test has generally been passing for months. In general, I don't think the right thing to do is skip a test if it's potentially a system-related issue, but I'll have this one added since it's very low risk.

**Comment by ryon-jensen** [comment]
Finally have a good run. I believe this is ready to land.

**Comment by daltonbohning** [comment]
@brianjmurrell Good with you?

**Review by brianjmurrell — APPROVED** [review]

**Pull request merged**
